control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+static std::vector<std::pair<cstring, MatchType>>
+getMatchFields(const IR::P4Table* table, ReferenceMap* refMap, TypeMap* typeMap) {
+  std::vector<std::pair<cstring, MatchType>> matchFields;
+
+  auto key = table->getKey();
+  if (!key) return matchFields;
+
+  for (auto keyElement : *key->keyElements) {
+    auto matchTypeDecl = refMap->getDeclaration(keyElement->matchType->path, true)
+                               ->to<IR::Declaration_ID>();
+    BUG_CHECK(matchTypeDecl != nullptr, "No declaration for match type '%1%'",
+                                        keyElement->matchType);
+    const auto matchTypeName = matchTypeDecl->name.name;
+
+    auto expr = keyElement->expression;
+    if (expr->is<IR::Slice>()) {
+      expr = expr->to<IR::Slice>()->e0;
+    }
+
+    MatchType matchType;
+    if (matchTypeName == P4CoreLibrary::instance.exactMatch.name) {
+      matchType = MatchField::EXACT;
+
+      // If this is an exact match on the result of an isValid() call, we
+      // desugar that to a "valid" match.
+      auto callToIsValid = tryCastToIsValidBuiltin(expr, refMap, typeMap);
+      if (callToIsValid != nullptr) {
+        expr = callToIsValid->appliedTo;
+        matchType = MatchField::VALID;
+      }
+    } else if (matchTypeName == P4CoreLibrary::instance.lpmMatch.name) {
+      matchType = MatchField::LPM;
+    } else if (matchTypeName == P4CoreLibrary::instance.ternaryMatch.name) {
+      matchType = MatchField::TERNARY;
+
+      // If this is a ternary match on the result of an isValid() call, desugar
+      // that to a ternary match against the '$valid$' field of the appropriate
+      // header instance.
+      // XXX(seth): This is baking in BMV2-specific stuff. We should remove this
+      // by performing the desugaring in a separate compiler pass.
+      auto callToIsValid = tryCastToIsValidBuiltin(expr, refMap, typeMap);
+      if (callToIsValid != nullptr) {
+        expr = new IR::Member(callToIsValid->appliedTo, "$valid$");
+      }
+    } else if (matchTypeName == P4V1::V1Model::instance.rangeMatchType.name) {
+      matchType = MatchField::RANGE;
+    } else {
+      ::error("Table '%1%': cannot represent match type '%2%' in PI",
+              table->externalName(), matchTypeName);
+      break;
+    }
+
+    HeaderFieldPath* path = HeaderFieldPath::from(expr, refMap, typeMap);
+    if (!path) {
+      ::error("Table '%1%': Match field '%2%' is too complicated "
+              "to represent in PI", table->externalName(), expr);
+      break;
+    }
+
+    // XXX(seth): If there's only one component in the path, then it represents
+    // a local variable. We need to support locals to support complex
+    // expressions in keys, because such expressions are desugared into a match
+    // against an intermediate local, but before implementing anything we need
+    // to decide how it makes sense to present something like that in PI.
+    if (!path->parent) {
```

Comment:
It *currently* is, but we can fix it. Let's leave this for now and fix it in a followup PR once we've had a chance to discuss how this should work in more detail.